### PR TITLE
[SourceKit] Add cancellation points for 'compile' requests

### PR DESCRIPTION
**Pull request by rintaro**
After each stage of the compilation pipeline before the code generation.
Sema, SILGen, SILOptimization, SILLowering, IRGen.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/CompileInstance.cpp [hunk 1]
```diff
@@ -325,7 +325,8 @@ bool CompileInstance::performCompile(
   if (CI->getDiags().hadAnyError())
```
Comment (ahoppen) [inline lib/IDE/CompileInstance.cpp:325]: Would it make sense to also check before `performSema`? That would catch cases where multiple compile request are queued up and a request is cancelled before it had a chance to get scheduled.

**Review by ahoppen — APPROVED** [review]
LGTM. One more idea for a cancellation point.

File: lib/IDE/CompileInstance.cpp (hunk 1 above)
Comment (rintaro) [reply lib/IDE/CompileInstance.cpp:325]: Good idea. 👍

**Comment by CodaFi** [comment]
These checks should probably shift into the request evaluator itself at some point.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux platform

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS platform

**Pull request merged**